pkg/vminetworkscontroller/vmi_controller.go

```diff
@@ -56,32 +58,58 @@ func (r *VirtualMachineInstanceReconciler) Reconcile(
 	defer cancel()
 	err := r.Client.Get(contextWithTimeout, request.NamespacedName, vmi)
 	if apierrors.IsNotFound(err) {
-		r.Log.Info("could not retrieve VMI - will cleanup its IPAMClaims")
-		if err := r.Cleanup(request.NamespacedName); err != nil {
-			return controllerruntime.Result{}, fmt.Errorf("error removing the IPAMClaims finalizer: %w", err)
-		}
-		return controllerruntime.Result{}, nil
+		vmi = nil
 	} else if err != nil {
 		return controllerruntime.Result{}, err
 	}
 
-	var ownerInfo metav1.OwnerReference
 	vm := &virtv1.VirtualMachine{}
 	contextWithTimeout, cancel = context.WithTimeout(ctx, time.Second)
 	defer cancel()
-	if err := r.Client.Get(contextWithTimeout, request.NamespacedName, vm); apierrors.IsNotFound(err) {
-		r.Log.Info("Corresponding VM not found", "vm", request.NamespacedName)
-		ownerInfo = metav1.OwnerReference{APIVersion: vmi.APIVersion, Kind: vmi.Kind, Name: vmi.Name, UID: vmi.UID}
-	} else if err == nil {
-		ownerInfo = metav1.OwnerReference{APIVersion: vm.APIVersion, Kind: vm.Kind, Name: vm.Name, UID: vm.UID}
-	} else {
+	hasVMOwner := false
+	if err := r.Client.Get(contextWithTimeout, request.NamespacedName, vm); err == nil {
+		hasVMOwner = true
+	} else if !apierrors.IsNotFound(err) {
 		return controllerruntime.Result{}, fmt.Errorf(
 			"error to get VMI %q corresponding VM: %w",
 			request.NamespacedName,
 			err,
 		)
 	}
 
+	if (hasVMOwner && vmi == nil && vm.DeletionTimestamp != nil) ||
+		(!hasVMOwner && (vmi == nil || (vmi.DeletionTimestamp != nil && len(vmi.Status.ActivePods) == 0))) {
```

Comment:
Agree, we are not going to understand what we are checking in a weeek or less.

Reply:
once we add helper here that is unit tested, maybe it makes 
https://github.com/kubevirt/ipam-extensions/pull/40#issuecomment-2242551362
unrequired ?
because the logic if to delete the finalizer from the claim (which what allows the claim to be deleted)
depends 99% solely on this code, this is the case to be precise:
"hasVMOwner && vmi == nil && vm.DeletionTimestamp != nil"

Just something to consider please

Reply:
I think you should add an e2e test w/ the e2e behavior, and I think you should add a unit test here to document what the criteria is for, and why the criteria is what it is.